src/include/migraphx/op/group_query_attention.hpp

```diff
@@ -0,0 +1,588 @@
+        const std::size_t present_buff_chunk_length =
+            present_buffer_sequence_length * head_size; // T x H
+
+        const std::size_t loop_len = batch_size * num_heads;
+        const float alpha = scale == 0.0f ? 1.0f / std::sqrt(static_cast<float>(head_size)) : scale;
+
+        par_for(loop_len, [&](const auto i) {
+            const std::size_t batch_index = i / num_heads;
+            const std::size_t head_index  = i % num_heads;
+            const std::size_t past_seqlen =
+                sequence_length == 1 ? seqlens_k[batch_index] : past_buffer_sequence_length;
+            const std::size_t past_chunk_length = past_seqlen * head_size;
+            const std::size_t total_seqlen      = seqlens_k[batch_index] + 1;
+
+            const std::size_t output_offset = i * sequence_length * present_buffer_sequence_length;
+            auto output                     = attention_probs + output_offset;
+
+            auto k = key + packed_batch_stride * batch_index +
+                     kv_input_chunk_length * (head_index / kv_num_heads_factor);
+            k = concat_state_chunk(past_key,
+                                   k,
+                                   present_key,
+                                   present_buff_chunk_length,
+                                   past_buff_chunk_length,
+                                   past_chunk_length,
+                                   kv_input_chunk_length,
+                                   is_prompt,
+                                   past_present_share_buffer,
+                                   i / kv_num_heads_factor);
+
+            // Calculate Q*K' + AttentionMask
+            //                     original                 transposed             each iteration
+            // A: Q                (B x N x) S x H          (B x N x) S x H        S x H
+            // B: K'               (B x N x) T x H          (B x N x) H x T        H x T
+            // C: attention_probs  (B x N x) S x T          (B x N x) S x T        S x T
+            auto q = query + packed_batch_stride * batch_index + q_input_chunk_length * head_index;
+            auto output_ptr = &(*output);
+            auto q_ptr      = &(*q);
+            auto k_ptr      = &(*k);
+            auto output_shape =
+                shape{dtype, {sequence_length, total_seqlen}, {present_buffer_sequence_length, 1}};
+            auto q_shape = shape{dtype, {sequence_length, head_size}, {head_size, 1}};
+            auto k_shape = shape{dtype, {head_size, total_seqlen}, {1, head_size}};
+            tensor_view<typename std::remove_pointer<decltype(output_ptr)>::type> cmat(output_shape,
+                                                                                       output_ptr);
+            tensor_view<typename std::remove_pointer<decltype(q_ptr)>::type> amat(q_shape, q_ptr);
+            tensor_view<typename std::remove_pointer<decltype(k_ptr)>::type> bmat(k_shape, k_ptr);
```

Comment:
There is a `make_view` function to deduce the type from the pointer: `auto bmat = make_view(k_shape, k_ptr);`